File: javascript/cluster_management/src/delete_multi_region_clusters.js
import { DSQLClient, DeleteClusterCommand, waitUntilClusterNotExists } from "@aws-sdk/client-dsql";

export async function deleteMultiRegionClusters(region1, cluster1Id, region2, cluster2Id) {

    const client1 = new DSQLClient({ region: region1 });
    const client2 = new DSQLClient({ region: region2 });

    try {
        const deleteClusterCommand1 = new DeleteClusterCommand({
            identifier: cluster1Id,
        });
        const response1 = await client1.send(deleteClusterCommand1);

        const deleteClusterCommand2 = new DeleteClusterCommand({
            identifier: cluster2Id,
        });
        const response2 = await client2.send(deleteClusterCommand2);

        console.log(`Waiting for cluster1 ${response1.identifier} to finish deletion`);
        await waitUntilClusterNotExists(
            {
                client: client1,
                maxWaitTime: 300 // Wait for 5 minutes
            },
            {
                identifier: response1.identifier
            }
        );
        console.log(`Cluster1 Id ${response1.identifier} is now deleted`);

        console.log(`Waiting for cluster2 ${response2.identifier} to finish deletion`);
        await waitUntilClusterNotExists(
            {
                client: client2,
                maxWaitTime: 300 // Wait for 5 minutes
            },
            {
                identifier: response2.identifier
            }
        );
        console.log(`Cluster2 Id ${response2.identifier} is now deleted`);
        return;
    } catch (error) {
        if (error.name === "ResourceNotFoundException") {
            console.log("Some or all Cluster ARNs not found or already deleted");
        } else {
            console.error("Unable to delete multi-region clusters: ", error.message);
        }
        throw error;
    }
}

async function main() {
    const region1 = process.env.CLUSTER_1_REGION || "us-east-1";
    const cluster1Id = process.env.CLUSTER_1_ID;
    const region2 = process.env.CLUSTER_2_REGION || "us-east-2";
    const cluster2Id = process.env.CLUSTER_2_ID;

    if (!cluster1_id || !cluster2_id) {
        console.error("Error: CLUSTER_1_ID and CLUSTER_2_ID environment variables must be set");
        process.exit(1);
    }

    await deleteMultiRegionClusters(region1, cluster1Id, region2, cluster2Id);
    console.log(`Deleted ${cluster1Id} in ${region1} and ${cluster2Id} in ${region2}`);
}

if (process.env.NODE_ENV !== 'test') {
    main();
}

File: java/cluster_management/src/main/java/org/example/DeleteMultiRegionClusters.java
package org.example;

import java.time.Duration;
import java.util.Optional;

import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.retries.api.BackoffStrategy;
import software.amazon.awssdk.services.dsql.DsqlClient;
import software.amazon.awssdk.services.dsql.DsqlClientBuilder;

public class DeleteMultiRegionClusters {

    public static void main(String[] args) {
        Region region1 = Region.of(System.getenv().getOrDefault("CLUSTER_1_REGION", "us-east-1"));
        String clusterId1 = Optional.ofNullable(System.getenv("CLUSTER_1_ID"))
                .orElseThrow(() -> new IllegalStateException("Expected CLUSTER_1_ID in environment"));
        Region region2 = Region.of(System.getenv().getOrDefault("CLUSTER_2_REGION", "us-east-1"));
        String clusterId2 = Optional.ofNullable(System.getenv("CLUSTER_2_ID"))
                .orElseThrow(() -> new IllegalStateException("Expected CLUSTER_2_ID in environment"));

        DsqlClientBuilder clientBuilder = DsqlClient.builder()
                .credentialsProvider(DefaultCredentialsProvider.builder().build());

        try (
            DsqlClient client1 = clientBuilder.region(region1).build();
            DsqlClient client2 = clientBuilder.region(region2).build()
        ) {
            example(client1, clusterId1, client2, clusterId2);
        }
    }

    public static void example(DsqlClient client1, String clusterId1, DsqlClient client2, String clusterId2) {
        System.out.printf("Deleting cluster %s in %s%n", clusterId1, client1.serviceClientConfiguration().region());
        client1.deleteCluster(r -> r.identifier(clusterId1));

        // cluster1 will stay in PENDING_DELETE until cluster2 is deleted.

        System.out.printf("Deleting cluster %s in %s%n", clusterId2, client2.serviceClientConfiguration().region());
        client2.deleteCluster(r -> r.identifier(clusterId2));

        // Now that both clusters have been marked for deletion they will transition
        // to DELETING state and finalize deletion.
        System.out.printf("Waiting for cluster %s to finish deletion%n", clusterId1);
        client1.waiter().waitUntilClusterNotExists(
                getCluster -> getCluster.identifier(clusterId1),
                config -> config.backoffStrategyV2(
                        BackoffStrategy.fixedDelayWithoutJitter(Duration.ofSeconds(10))
                ).waitTimeout(Duration.ofMinutes(5))
        );

        System.out.printf("Waiting for cluster %s to finish deletion%n", clusterId2);
        client2.waiter().waitUntilClusterNotExists(
                getCluster -> getCluster.identifier(clusterId2),
                config -> config.backoffStrategyV2(
                        BackoffStrategy.fixedDelayWithoutJitter(Duration.ofSeconds(10))
                ).waitTimeout(Duration.ofMinutes(5))
        );

        System.out.printf("Deleted %s in %s and %s in %s%n",
                clusterId1, client1.serviceClientConfiguration().region(),
                clusterId2, client2.serviceClientConfiguration().region());
    }
}
